### PR TITLE
AArch64 timer.

### PR DESCRIPTION
**Pull request by pj1031999**
Enable timer on AArch64.

It depends on changes from #769, #770, #771, #772, #787, #788.

### CODE REVIEW

File: include/aarch64/bcm2835reg.h [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+/*
+ * Reference: BCM2835 ARM Periperhals
+ *
+ * 	http://www.raspberrypi.org/wp-content/uploads/2012/02/BCM2835-ARM-Peripherals.pdf
+ */
+
+#ifndef _BCM2835REG_H_
+#define _BCM2835REG_H_
+
+#define BCM3836
```
Comment (cahirwpz) [inline include/aarch64/bcm2835reg.h:41]: Typo?

Comment (pj1031999) [reply include/aarch64/bcm2835reg.h]: Fixed.

File: sys/aarch64/rootdev.c [hunk 2]
```diff
@@ -72,9 +170,10 @@ static void rootdev_intr_teardown(device_t *dev, intr_handler_t *handler) {
   intr_event_remove_handler(handler);
 }
 
-/* Read 32 bit pending register located at va and run handlers. */
-static void bcm2835_intr_handle(vaddr_t irqpendr, intr_event_t *events) {
-  uint32_t pending = *(uint32_t *)irqpendr;
+/* Read 32 bit pending register located at irqpendr and run handlers. */
```
Comment (ThomasStachovsky) [inline sys/aarch64/rootdev.c:173]: There's no ```irqpendr``` here.

File: sys/aarch64/timer.c [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+#define CNTCTL_ENABLE 1
+
+typedef struct arm_timer_state {
+  intr_handler_t intr_handler;
+  timer_t timer;
+  uint64_t step;
+} arm_timer_state_t;
+
+static int arm_timer_start(timer_t *tm, unsigned flags, const bintime_t start,
```
Comment (ThomasStachovsky) [inline sys/aarch64/timer.c:18]: ```flags``` and ```start``` are unused. They are required by definition of ```tm_start_t```, but maybe mark them as ```__unused```?

**Review by ThomasStachovsky — CHANGES_REQUESTED** [review]
I left a few comments, I believe there's a mistake in rootdev.c in lines 119 and 142.

File: sys/aarch64/timer.c [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    uint64_t count = READ_SPECIALREG(cntpct_el0);
+    WRITE_SPECIALREG(cntp_cval_el0, count + state->step);
+    WRITE_SPECIALREG(cntp_ctl_el0, CNTCTL_ENABLE);
+  }
+
+  return 0;
+}
+
+static int arm_timer_stop(timer_t *tm) {
+  WRITE_SPECIALREG(cntp_ctl_el0, 0);
```
[the remark this answers is not in the document]
Comment (pj1031999) [reply sys/aarch64/timer.c]: done

File: sys/aarch64/timer.c [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  state->intr_handler =
+    INTR_HANDLER_INIT(arm_timer_intr, NULL, dev, "ARM CPU timer", 0);
+
+  timer_t *tm = &state->timer;
+
+  uint64_t freq = READ_SPECIALREG(cntfrq_el0);
+  tm->tm_frequency = freq;
+  tm->tm_min_period = HZ2BT(freq);
+  /* TODO(pj): Document what this value means. */
+  tm->tm_max_period = bintime_mul(HZ2BT(freq), 1LL << 30);
```
[the remark this answers is not in the document]
Comment (pj1031999) [reply sys/aarch64/timer.c]: done

File: sys/aarch64/rootdev.c [hunk 6]
```diff
@@ -38,6 +43,11 @@ static int rootdev_bs_map(bus_addr_t addr, bus_size_t size,
     *handle_p = rootdev_local_handle;
     return 0;
   }
+  if (addr >= BCM2835_PERIPHERALS_BUS_TO_PHYS(BCM2835_ARM_BASE) &&
+      addr < BCM2835_PERIPHERALS_BUS_TO_PHYS(BCM2835_ARM_BASE) + PAGESIZE) {
```
[the remark this answers is not in the document]
Comment (pj1031999) [reply sys/aarch64/rootdev.c]: done

File: sys/aarch64/rootdev.c [hunk 7]
```diff
@@ -59,6 +69,94 @@ static bus_space_t *rootdev_bus_space = &(bus_space_t){
...
+  bus_space_write_4(rootdev_bus_space, rootdev_arm_base,
+                    BCM2835_ARMICU_OFFSET + offset, reg & (~(1 << irq)));
+}
+
+static void rootdev_enable_irq(intr_event_t *ie) {
+  int irq = ie->ie_irq;
+  assert(irq < NIRQ);
+
+  /* Enable local IRQ. */
+  if (irq < BCM2835_NIRQ) {
```
[the remark this answers is not in the document]
Comment (pj1031999) [reply sys/aarch64/rootdev.c]: fixed

File: sys/aarch64/rootdev.c [hunk 8]
```diff
@@ -59,6 +69,94 @@ static bus_space_t *rootdev_bus_space = &(bus_space_t){
...
+  /* Enable base IRQ. */
+  enable_gpu_irq(irq - BCM2835_INT_BASICBASE, BCM2835_INTC_IRQBPENDING);
+}
+
+static void rootdev_disable_irq(intr_event_t *ie) {
+  int irq = ie->ie_irq;
+  assert(irq < NIRQ);
+
+  /* disable local IRQ. */
+  if (irq < BCM2835_NIRQ) {
```
[the remark this answers is not in the document]
Comment (pj1031999) [reply sys/aarch64/rootdev.c]: fixed

File: sys/aarch64/timer.c (hunk 3 above)
Comment (pj1031999) [reply sys/aarch64/timer.c]: done

File: sys/aarch64/rootdev.c (hunk 2 above)
Comment (pj1031999) [reply sys/aarch64/rootdev.c]: Fixed by @cahirwpz.

**Review by cahirwpz — APPROVED** [review]
LGTM

**Pull request merged**